Check in "Rust for Embedded C Programmers" (#2474)
lowRISC/opentitan: doc/ug/rust_for_c.md

*   `cargo doc --open` builds crate documentation and then opens it in the browser.
*   `cargo fmt` runs the Rust formatter[^9].

Also, the contents of the `RUSTFLAGS` environment variable are passed to `rustc`, as a mechanism for injecting flags.

The Rust standard library, like libc, is smaller in embedded environments.
The standard library consists of three crates: `core`[^10], `alloc`, and `std`.
`core`, sometimes called libcore, is all of the fundamental definitions, which don't rely on operating system support.
Nothing in `core` can heap-allocate.
`alloc` does not require OS support, but does require `malloc` and `free` symbols.
`std` is `core` + `alloc`, as well as OS APIs like file and thread support.
The `#[no_std]` pragma disables `std` and `alloc`, leaving behind `core`.
Throughout this document, we will only use `core` types, though we may refer to them through the `std` namespace (they're aliases).
That is, we may refer to `std::mem::drop`, though in `#[no_std]` code it must be named `core::mem::drop`.

rustc has a number of flags. The most salient of these are:

*   `--emit asm` and `--emit llvm-ir`, useful for inspecting compiler output.

> *   `--target`, roughly equivalent to Clang's `-march`.

Author: Done. I guess this shows how little I know about Clang's interface... =P

> *   `--target`, roughly equivalent to Clang's `-march`.

Author: Shows that I don't know my Clang flags... fixed.

  a: u32,
  b: u8,
}
```

This makes it possible to use `==` on `MyStruct` values, which will just perform field-wise equality.
The same can be done for ordering operations like `<` and `>=`: `#[derive(PartialOrd, Ord)]` will define comparison functions that compare structs lexicographically.

#### Enums and Unions

Much like C, Rust has enumerations for describing a type with a fixed set of values:
```rust
enum MyEnum {
    Banana, Apple, Pineapple,
}
```

Unlike C, though, `MyEnum` is a real type, and not just an alias for an integer type.
Also unlike C, the enum's variants are not dumped into the global namespace, and must instead be accessed through the enum type: `MyEnum::Banana`.
Note that, unlike structs, the variants of an enum are automatically public.

While Rust does represent enum values with integers (these integers are called _discriminants_), the way they're laid out is unspecified.
To get an enum whose discriminants are allocated like in C, we can employ a `repr` attribute:
```rust
#[repr(C)]
enum MyCEnum {
  Banana = 0,
  Apple = 5,
  Pineapple = 7,
}
```
Unlike C, though, Rust will only guarantee discriminant values that are explicitly written down.

> Such enums may be safely cast into integer types (like `MyCEnum::Apple as u32`), but not back: the compiler always assumes that the underlying value of a `MyCEnum` is `0`, `5`, or `7`, and violating this constraint is UB[^23].

Author: Done.

unsafe { my_union.bar }  // I assert that bar is the active variant.
```
Unions also have restrictions on what types can be used as variants, due to concerns about destructors.

Since unions are so useful in C, but utterly unsafe, Rust provides built-in _tagged unions_[^25], which are accessed through the `enum` syntax:
```rust
enum MyEnum {
  FooVariant { foo: i32 },
  BarVariant(u8),
}
```

Tagged-union enum variants use the same syntax as Rust structs; enum consists of a tag value (the discriminant) big enough to distinguish between all the variants, and a compiler-tracked union of variants.
However, effective use of such enums requires _pattern matching_, which requires its own section to explain.
We'll see these enums again when we discuss patterns.

Just like with structs, `#[derive]` can be used on enums to define comparison operators, which are defined analogously to the struct case.

#### Arrays {#arrays}

Rust arrays are just C arrays: inline storage of a compile-time-known number of values.
`T[N]` in C is spelled `[T; N]` in Rust.
Arrays are created with `[a, b, c]` syntax, and an array with lots of copies of the same value can be created using `[0x55u8; 1024]`[^26].

> A mutli-dimensional array can be declared as an array of arrays: `[[T; N]; M]`.

Author: Done.

Array elements can be accessed with `x[index]`, exactly like in C.
Note, however, that Rust automatically inserts bounds checks around every array access[^27]; failing the bounds check will trigger a panic in the program.
Unsafe Rust can be used to cheat the bounds check when it is known (to the programmer, not rustc!) to be unnecessary to perform it, but when it is performance critical to elide it.

Rust arrays are "real" types, unlike in C. They can be passed by value intofunctions, and returned by value from functions.

> They also don't decay into points when passed into functions.

Author: Done.

Mutable globals are particularly dangerous, since they can be a source of data races in multicore systems.
As such, reading or writing to a mutable global, or creating a reference to one, requires Unsafe Rust.

#### Functions {#functions}

In C and Rust, functions are the most important syntactic construct. Rust declares functions like so:
```rust
fn my_function(x: u32, y: *mut u32) -> bool {
  // Function body.
}
```
The return type, which follows the `->` token, can be omitted when it is `()`, Rust's `void` type.
Functions are called with the usual `foo(a, b, c)` syntax.

The body of a function consists of a list of statements, potentially ending in an expression; that expression is the function's return value (no `return` keyword needed).
If the expression is missing, then `()` is assumed as the return type. Items can be mixed in with the statements, which are local to their current scope but visible in all of it.

Rust functions can be marked as `unsafe fn`.
This means the function cannot be called normally, and must instead be called using Unsafe Rust.

> The body of an `unsafe fn` behaves like an `unsafe` block; we'll go more into this when we discuss Unsafe Rust in detail.

Author: Yes, I believe this is true.

> The body of an `unsafe fn` behaves like an `unsafe` block; we'll go more into this when we discuss Unsafe Rust in detail.

Author: That sounds right. I remember it being pretty favorable but I haven't heard about it lately.

The block will then evaluate to the final expression. For example:
```rust
let foo = {
  let bar = 5;
  bar ^ 2
};
```
Blocks are like local functions that execute immediately, and can be useful for constraining the scope of variables.

#### Conditionals: `if` and `match` {#conditionals-if-and-match}

Rust's `if` expression is, syntactically, similar to C's. The full syntax is
```rust
if cond1 {
  // ...
} else if cond2 {
  // ...
} else {
  // ...
}
```

Conditions must be expressions that evaluate to `bool`.
A conditional may have zero or more `else if` clauses, and the `else` clause is optional.
Because of this, Rust does not need (and, thus, does not have) a ternary operator:
```rust
let x = if c { a } else { b };
```

Braces are required on `if` expressions in Rust.

`if` expressions evaluate to the value of the block that winds up getting executed.
As a consequence, all blocks must have the same type[^50].
For example, the following won't compile, because of an errant semicolon:
```rust
if cond() {
  my_int(4)   // Type is i32.
} else {
  my_int(7);  // Type is (), due to the ;
}
```

> In general, it is a good idea to end all expressions with a semicolon, unless its value is needed.

Author: Hmm... I'm not sure. The point I wanted to make was "when you use a conditional, all the values get unified; to avoid unification errors, remember to include a semicolon". I think this advice applies less strongly to blocks that are actively being used as expressions, where the semicolon is actually totally unimportant unless you assign the block to a `let`.

Maybe it should be reworded though, since the semicolons are _required_ for expressions prior to the final one?

> In general, it is a good idea to end all expressions with a semicolon, unless its value is needed.

Reviewer: If I understand you, your contention is that this is the main place the final semicolon (or not) will bite you, and that the end of a function's block is less likely because of explicit `return`. 

Ok, I can get behind that. Maybe do a forward reference to `potential ending expression` from `Blocks` to this paragraph instead then?

> In general, it is a good idea to end all expressions with a semicolon, unless its value is needed.

Author: Done.

#### Enums and Unions

Much like C, Rust has enumerations for describing a type with a fixed set of values:
```rust
enum MyEnum {
    Banana, Apple, Pineapple,
}
```

Unlike C, though, `MyEnum` is a real type, and not just an alias for an integer type.
Also unlike C, the enum's variants are not dumped into the global namespace, and must instead be accessed through the enum type: `MyEnum::Banana`.
Note that, unlike structs, the variants of an enum are automatically public.

While Rust does represent enum values with integers (these integers are called _discriminants_), the way they're laid out is unspecified.
To get an enum whose discriminants are allocated like in C, we can employ a `repr` attribute:
```rust
#[repr(C)]
enum MyCEnum {
  Banana = 0,
  Apple = 5,
  Pineapple = 7,
}
```
Unlike C, though, Rust will only guarantee discriminant values that are explicitly written down.
Such enums may be safely cast into integer types (like `MyCEnum::Apple as u32`), but not back: the compiler always assumes that the underlying value of a `MyCEnum` is `0`, `5`, or `7`, and violating this constraint is Undefined Behavior (UB)[^23].

> If we want to require that an enum is an exact integer width, we can use `#[repr(uN)]`, where `uN` is a type like `u16`.

Author: Done.

#### Constants and Globals {#constants-and-globals}

Rust has dedicated syntax for compile-time constants, which serve the same purpose as `#define`d constants do in C.
Their syntax is
```rust
const MY_CONSTANT: u32 = 0x42;
```
The type is required here, and the right-hand side must be a _constant expression_, which is roughly any combination of literals, numeric operators, and `const fn`s (more on those in a moment).

Constants do not exist at runtime.
They can be thought of best as fixed expressions that get copy+pasted into wherever they get used, similar to `#define`s and `enum` declarators in C.
It is possible to take the address of a constant, but it is not guaranteed to be consistent.[^42]

Globals look like constants, but with the keyword `static`[^43]:
```rust
static MY_GLOBAL: u8 = 0x00;
static mut MY_MUTABLE_GLOBAL: Foo = Foo::new();
```
Globals are guaranteed to live in `.rodata`, `.data`, or `.bss`, depending on their mutability and initialization.
Unlike constants, they have unique addresses, but as with constants, they must be initialized with constant expressions.[^44]

> Mutable globals are particularly dangerous, since they can be a source of data races in multicore systems.

Author: Oooh, good point. I don't think those are data races in the traditional sense, but I think you can probably observe nonatomicity at the compiler level (e.g. pre-loading into registers, speculative reads, etc).

As such, reading or writing to a mutable global, or creating a reference to one, requires Unsafe Rust.

#### Functions {#functions}

In C and Rust, functions are the most important syntactic construct. Rust declares functions like so:
```rust
fn my_function(x: u32, y: *mut u32) -> bool {
  // Function body.
}
```

> The return type, which follows the `->` token, can be omitted when it is `()`, Rust's `void` type.

Author: `()` is already named at this point, but I'll add it for redundancy.

`loop` is unique to Rust; it is simply an infinite loop[^52]:
```rust
loop { /* Stuff. */ }
```
Because infinite loops can never end, the type of the `loop` expression (without a `break` in it!) is `!`, since any code after the loop is dead.
Having an unconditional infinite loop allows Rust to perform better type and lifetime analysis on loops.
Under the hood, all of Rust's control flow is implemented in terms of `loop`, `match`, and `break`.

#### Control Flow {#control-flow}

Rust has `return`, `break`, and `continue`, which have their usual meanings from C.
They are also expressions, and, much like `loop {}`, have type `!` because all code that follows them is never executed (since they yank control flow).

`return x` exits a function early with the value `x`.
`return` is just syntax for `return ()`.
`break` and `continue` do the usual things to loops.

All kinds of loops may be annotated with labels (the only place where Rust allows labels):
```rust
'a: loop { .. }
```
`break` and `continue` may be used with those labels (e.g. `break 'a`), which will break or continue the enclosing loop with that label (rather than the closest enclosing loop).

> While C lacks this feature, most languages without `goto` have it.

Author: A fairly old RFC'd feature is being able to annotate any block with `'a:`, which can be broken-with-value to. (This is sort of like call/cc but with compile-time continuation locations.)

Also, some care must be taken with what types are sent over the boundary.
See [https://doc.rust-lang.org/reference/items/external-blocks.html](https://doc.rust-lang.org/reference/items/external-blocks.html) for more details.

### Analogues of other functionality {#analogues-of-other-functionality}

#### Volatile {#volatile}

Rust does not have a `volatile` qualifier.
Instead, volatile reads can be performed using the `read_volatile()`[^55] and `write_volatile()`[^56] methods on pointers, which behave exactly like volatile pointer dereference in C.

Note that these methods work on types wider than the architecture's volatile loads and stores, which will expand into a series of volatile accesses, so beware.
The same caveat applies in C: `volatile uint64_t` will emit multiple accesses on a 32-bit machine.

#### Inline Assembly {#inline-assembly}

Rust does not quite support inline assembly yet.
Clang's inline assembly syntax is available behind the unstable macro `asm!()`, which will eventually be replaced with a Rust-specific syntax that better integrates with the language.
`global_asm!()` is the same, but usable in global scope, for defining whole functions.

> Naked functions can be created using `#[naked]`. See [https://doc.rust-lang.org/1.8.0/book/inline-assembly.html](https://doc.rust-lang.org/1.8.0/book/inline-assembly.html).

Author: The ironic part is that I contributed to that RFC...

[Note that this syntax is currently in the process of being redesigned and stablized.](https://blog.rust-lang.org/inside-rust/2020/06/08/new-inline-asm.html)

#### Bit Casting {#bit-casting}

> Rust provides a type systemrap-door for bitcasting any type to any other type of the same size:

Author: Done.

```c
// `handle` is a managed resource to a peripheral, that should be
// destroyed to signal to the hardware that the resource is not in use.
my_handle_t handle = new_handle(0x40000);
use_for_scheduling(handle);  // Does something with `handle` and destroys it.
// ... 200 lines of scheduler code later ...
use_for_scheduling(handle);  // Oops double free.
```
Double-free and use-after-free are a common source of crashes and security vulnerabilities in C code.
Let's see what happens if we were to try this in Rust.

Consider the equivalent code written in Rust:
```rust
let handle = new_handle(0x40000);
use_for_scheduling(handle);
// ...
use_for_scheduling(handle);
```

If you then attempt to compile this code, you get an error:
```rust
error[E0382]: use of moved value: `handle`
  --> src/main.rs:10:24
   |
7  |     let handle = new_handle(0x40000);
   |         ------ move occurs because `handle` has type `Handle`,
   |                which does not implement the `Copy` trait
8  |     use_for_scheduling(handle);
   |                        ------ value moved here
9  |     // ...
10 |     use_for_scheduling(handle);
   |                        ^^^^^^ value used here after move
```

> Use-after-free errors (and double-free errors) are impossible in Rust.

Author: Done.

#### Operations on References {#operations-on-references}

Rust references behave much more like scalar values than like pointers (borrow-checking aside).
Because it is statically known that every reference, at all times, points to a valid, initialized value of type `T`, explicit dereferencing is elided most of the time (though, when necessary, they can be dereferenced: `*x` is an lvalue that can be assigned to).

Rust does not have a `->` operator, but, for `x: &T`, the dot operator behaves as if `x` were a `T`.
If `field` is a field of `T`, for example, `x.field` is the lvalue of `field` (which would be spelled `x->field` in C).
This applies even for heavily nested rerences: the dot operator on `&&&T` will trigger three memory lookups.
This is called the "auto-deref" behavior.

Equality of references is defined as equality of underlying values: `x == y`, for `x: &T` and `y: &T`, becomes `*x == *y`.
Pointer equality is still possible with `std::ptr::eq(x, y)`.
References can be coerced into raw pointers: `x as *const T`, and compared directly.

### Methods {#methods}

> While Rust is not an object-oriented language, it does provide a mechanism for namespacing functns under types: `impl` (for implementation) blocks.

Author: Done.

```rust
impl MyStruct<'_> { .. }
```
As we've already seen, many primitive types have methods, too; these are defined in special `impl` blocks in the standard library.

### Slices and `for` {#slices-and-for}

References also do not allow for pointer arithmetic, so a `&u32` cannot be used to point to a buffer of words.
Static buffers can be passed around as arrays, like `&[u32; 1024]`, but often we want to pass a pointer to contiguous memory of a runtime-known value.
_Slices_ are Rust's solution to pointer+length.

A slice of `T` is the type `[T]`; this type is most like a "flexible array member" in C:
```rust
struct slice {
  size_t len;
  T values[];
}
```
Then, a `slice*` would point to a length followed by that many `T`s; it can't reasonably exist except behind a pointer.
Similarly, `[T]` is what Rust calls a dynamically sized type[^70], which needs to exist behind a reference: it is much more common to see `&[T]` and `&mut [T]`.

However, Rust still differs from the C version: `&[T]` is a _fat pointer_, being two words wide.
It essentially looks like this:
```rust
struct Slice {
  len: usize,
  values: *const T,
}
```

> A reference to slice otherwise works like an array reference: `&x[n]` extracts a reference to the `n`th element in the slice (with bounds checking), `x[n] = y` assigns to it.

Author: Done.

}
```
*   Enum patterns are probably the most important kind of pattern, and are what we saw in the `match` statement with `Option` above.
    They're very similar to struct patterns, except that instead of always succeeding, they check that the enum discriminant is the one named in the pattern.
```rust
enum MyEnum { A, B{u32), C { a: i32, b: i32 }, }
match my_enum {
  MyEnum::A => { ... }  // Match for variant `A`.
  MyEnum::B(7) => { ... }  // Match for variant `B`, with 7 as the value inside.
  MyEnum::B(x) => { ... }  // Match for variant `B`, binding the value inside to
                           // `x`.
  MyEnum::C { a: 7, .. } => { ... }  // Match for variant `C`, with 7 as the
                                     // value in `a` and all other fields ignored.
  MyEnum::C { b, .. } => { ... }  // Match for variant `C`, binding b to b.
}
```
A complete treatment of the pattern syntax can be found at [https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html](https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html).

A `match` expression will evaluate each pattern against a value until one matches, in order; the compiler will warn about patterns which are unreachable[^87].

> The compiler will also ensure that every value will get matched with one of the match arms, either because every case is covered (e.g., every enum variant is present) or an _irrefutable pattern_ is present (i.e., a patte which matches all values). `_`, `foo`, `(a, _)`, and `MyStruct { a, .. }` are all examples of irrefutable patterns.

Reviewer 2: ```suggestion
The compiler will also ensure that every value will get matched with one of the match arms, either because every case is covered (e.g., every enum variant is present) or an _irrefutable pattern_ is present (i.e., a pattern which matches all values). `_`, `foo`, `(a, _)`, and `MyStruct { a, .. }` are all examples of irrefutable patterns.
```

Author: Done.

    If any of the patterns match, the arm executes.
*   _Match guards_ give you a shorthand for conditioning an arm on some expression: `Some(foo) if foo.has_condition() => ...,`.

Also, the standard library provides the `matches!()` macro as a shorthand for the following common match expression:
```rust
match expr {
  <some_complex_match_arm> => true,
  _ => false,
}
// ... can be replaced with ...
matches!(expr, some_complex_match_arm)
```
`matches!` supports multi-match and match guards as well.

Irrefutable patterns can be used with normal variable declaration.
The syntax `let x = ...;` actually uses a pattern: `x` is a pattern.
Other irrefutable patterns can be used there:
```rust
let (a, b) = ...;  // Destructure a tuple, rather than using clunky `.0` and `.1` field names.
let Foo { foo, bar, baz } = ...;  // Destructure a struct, to access its fields directly.
let 42 = ...;  // Not allowed: `42` is not an irrefutable pattern.
```

Special variants of `if` and `while` exist to take advantage of patterns, too:
```rust
if let Some(x) = my_option {
  // If the pattern succeeds, the body will be executed, and `x` will be bound

>   // to the value inside the Option.

Reviewer 2: I think it can be helpful to show x being used inside this block, so it is clear x is only bound inside the case for the successful block

Author: Done.

} else {
  // Else block is optional.
}

while let Some(x) = some_func() {
  // Loop terminates once the pattern match fails. Again, `x` is bound
  // to the value inside the Option.
}
```
Unlike normal `let` statements, `if let` and `while let` expressions are meant to be used with refutable patterns.

In general, almost every place where a value is bound can be an irrefutable pattern, such as function parameters and `for` loop variables:
```rust
fn get_first((x, _): (u32, u32)) -> u32 { x }

for (k, v) in my_key_values { .. }
```

### Traits {#traits}

Traits are Rust's core code-reuse abstraction.
Rust traits are like interfaces in other languages: a list of methods that a type must implement.

> Traits themselves, however, are _not _types.

Reviewer: ```suggestion
Traits themselves, however, are _not_ types.
```

Author: This one we can blame on the conversion tool I used. =P

A very simple trait from the standard library is `Clone`[^88]:
```rust
trait Clone {
  fn clone(&self) -> Self;
}
```

A type satisfying `Clone`'s interface (in Rust parlance, "implements `Clone`") has a `clone` method with the given signature, which returns a duplicate of `self`.
To implement a trait, you use a slightly funny `impl` syntax[^89]:
```rust
impl Clone for MyType {
  fn clone(&self) -> Self { /* implementation */ }
}
```

This gives us a consistent way to spell "I want a duplicate of this value".
The standard library provides traits for a number of similar operations, such as `Default`[^90], for providing a default value, `PartialEq`[^91] and `Eq`, for equality, `PartialOrd`[^92] and `Ord`, for ordering, and `Hash`[^93], for non-cryptographic hashing.

> The above traits are special in that they have trivial implementations for a struct or enum, assuming that all fields of stat struct or enum implement it.

Reviewer: ```suggestion
The above traits are special in that they have trivial implementations for a struct or enum, assuming that all fields of that struct or enum implement it.
```

Author: Done.

> The above traits are special in that they have trivial implementations for a struct or enum, assuming that all fields of stat struct or enum implement it.

Reviewer 2: ```suggestion
The above traits are special in that they have trivial implementations for a struct or enum, assuming that all fields of that struct or enum implement it.
```

Author: Sam already caught this one. =)

This is separate from marking a trait as `unsafe`, and it is not necessary to mark a trait as `unsafe` for it to have `unsafe` methods.

#### Auto Traits

Auto traits are a compiler mechanism for automatically implementing certain traits; in the standard library's source code, this shows up as `auto trait Foo {}` (though this syntax is unavailable for normal libraries).
Auto traits are implemented automatically for a struct or enum type if all of its fields also implement that trait, and are used for exposing transitive properties to the trait system.
For example, `Send` and `Sync` are auto traits; a number of other marker traits[^99] are also auto traits.

### Generics {#generics}

_Generic programming_ is writing source code that can be compiled for many types.
Generics are one of Rust's core features, which enable polymorphic[^100] static dispatch.

Functions can be made generic by introducing _type parameters_, using a syntax similar to explicit lifetimes[^101]:
```rust
fn identity<T>(x: T) -> T {
  x
}
```
This function accepts a value of any type and immediately returns it.
It can then be called like this: `identity::<i32>(42)`[^102].

> Using a generic function with all of its type parameters filled in causes it to be _instantiated_ (or _monomorphized)_, resulting in code being generated for it.

Reviewer: ```suggestion
Using a generic function with all of its type parameters filled in causes it to be _instantiated_ (or _monomorphized_), resulting in code being generated for it.
```

Author: Done.

If it were applied to `f`, `MyClosure` would become
```rust
struct MyClosure<'a> {
  x: X,
  y: Y,
}
```

Rust does not cleanly support mixing capture-by-move and capture-by-reference, but it is possible to mix them by capturing references by move:
```rust
let x = ...;
let y = ...;
let x_ref = &x;
let f = move |arg| {
  x_ref.do_thing(arg);  // Capture x_ref by move, aka capture x by shared ref.
  y.do_mut_thing(arg);  // Capture y by move.
};
```
The distinction between capture-by-move and capture-by-ref is mostly irrelevant for `Copy` types.

To be polymorphic over different closure types, we use the special `Fn`, `FnMut`, and `FnOnce` traits.
These represent functions that can be called by shared reference, unique reference, or by move.
Closures that only capture shared references implement all three; closures that capture by unique reference implement only the latter two, and closures that capture by move implement only the last one[^115].
Function pointers, function items[^116], and closures that don't capture also implement all three, and can all be converted to function pointers.

> These traits usspecial syntax similar to function pointers[^117].

Reviewer: ```suggestion
These traits use special syntax similar to function pointers[^117].
```

Author: Done.

```
let x = my_thing.foo()?.bar()?.baz()?;
```
See [https://doc.rust-lang.org/std/result/index.html](https://doc.rust-lang.org/std/result/index.html) for more `Result` operations.

### `for` Revisited: Iterators {#for-revisited-iterators}

An _iterator_ is any type that implements the `Iterator` trait, which looks like this:
```rust
trait Iterator {
  type Item;
  fn next(&mut self) -> Option<Self::Item>;
}
```

An iterator produces a sequence of `Option<Item>` values; the `next()` method allows an iterator to advance some internal state and produce the next value in the sequence.

For example, a very simple iterator simply produces every nonnegative integer value, in sequence:
```rust
struct Counter { state: u64 }
impl Iterator for Counter {
  type Item = u64;
  fn next(&mut self) -> Option<u64> {
    let current = self.state;
    self.state += 1;
    Some(current)
  }
}
```

This iterator will produce values forever: it always returns `Some`.
An iterator that eventually produces `None`, and then forever more returns `None`, is called "fused"[^121].

> Some iterators may start returning `Some`, but most Rust constructs treat all iterators as if they were fused.

Reviewer: ```suggestion
Some iterators may start returning `Some` after returning `None`, but most Rust constructs treat all iterators as if they are fused.
```

Author: Done.

     Applies a predicate to each element; if the predicate returns `false`, it is skipped by `next()`.

A number of other traits can enhance the properties of an iterator, enabling further methods: `ExactSizeIterator` iterators produce a known, fixed number of values; `DoubleEndedIterators` can have elements pulled from both the front, and the back of the sequence.
While many of the operations above have naive implementations in terms of `next`, standard library iterators will override them when a more efficient algorithm is available.
In general, iterators can produce very efficient code similar to that emitted by `while` loops, but care should be taken when using especially complex chains of combinators.

See [https://doc.rust-lang.org/std/iter/trait.Iterator.html](https://doc.rust-lang.org/std/iter/trait.Iterator.html) and [https://doc.rust-lang.org/std/iter](https://doc.rust-lang.org/std/iter/) for full details on available operations.

### Modules and Crate Layout {#modules-and-crate-layout}

Each Rust crate is (from the compiler's perspective) given a unique, single-identifier name.
This name is used to namespace a crate's symbols[^122].
`core` and `std` are crates.

> Each crate is rooted  a `lib.rs` or `main.rs` file, depending on whether it is a library or a binary.

Reviewer: ```suggestion
Each crate is rooted in a `lib.rs` or `main.rs` file, depending on whether it is a library or a binary.
```

Author: Done.

### Unsafe Rust {#unsafe-rust}

Unsafe Rust is a dialect of Rust that is denoted by the keyword `unsafe`: `unsafe` blocks, unsafe functions, unsafe traits.
Importantly, all of these actions require uttering the keyword `unsafe`, so that they can be easily detected in code review.
Code inside unsafe blocks says to the  reader that the programmer has checked subtle safety guarantees that the compiler cannot on its own.

Unsafe Rust fundamentally works by "turning off" certain checks the compiler normally enforces, whenever inside a `unsafe { ... }` block expression or the body of an `unsafe fn`.
The things that Unsafe Rust can do that Safe Rust cannot are:
*   Call `unsafe` functions.
*   Dereference raw pointers.
*   Mutate global state through a mutable `static`.
*   Read union fields.
*   Call the `asm!` macro.

Additionally, `unsafe impl`s may implement `unsafe trait`s, but don't need to be inside an `unsafe` block.

The canonical reference is the Rustonomicon[^127], a non-normative document describing common uses for Unsafe Rust.

> It is mostly required reading (mostly the first half) for embedded programming.

Reviewer: ```suggestion
It is required reading (mostly the first half) for embedded programming.
```

Author: Done.

In general, a few rules of thumb are useful:
*   Every `unsafe fn` should declare, in documentation, what invariants it assumes that the caller will uphold, and what state it will leave the machine in.
    For example, `<[T]>::get_unchecked(n)` elides the bounds check for the indexing operation, and it is up to the caller to uphold it instead.
*   Every time `unsafe` code calls into a non-`unsafe` function, it must ensure that no violated invariants, which could trigger Undefined Behavior, are observable in that safe code.
    For example, if we have a type that maintains the invariant that `len > 0`, and we temporarily set it to `0` during an `unsafe` block, it must be restored to `> 0` before any safe methods can be called on that type[^128].
*   Unsafe code should be kept to the absolute minimum, and wrapped in safe interfaces that assert invariants, either through static type-system guarantees or through runtime checks.
    Every line of unsafe code is a place where the engineering cost of Rust's guarantees are wasted.

> In other words, Safe Rust is able to freely assume it cannot trigger Undefined Behavior, at all times, no matter what.

Reviewer: I'm not sure this sentence is true.

Author: So, my read has always been "if I program contains `#![forbid(unsafe_code)]`, _and_ none of its dependencies can invoke UB under any circumstance, then that program may be assumed to be well-defined for all inputs", which is what I'm trying to get across with this sentence. (I don't think Rust considers stack overflow or OOM to be UB in the same way C++ does, at any rate...)

Is this actually false under some circumstances? I know that `loop {}` emits LLVM UB, but that's a long-standing rustc bug, and certainly not Rust UB.

Reviewer: I think you're trying to tie yourself into knots by starting with a strong statement and then having quite a few implicit caveats. 

I would tone down the statement rather than adding more footnotes. You want something like "Safe Rust can assume that Rust's overall aliasing, ownership, and representation invariants are preserved at all times."

Author: Ok. I incorporated your suggested language. Do you think my new phrasing gets the point across better?

Reviewer: Yep.

[^63]: These days, it's a subgraph of the control flow graph.

[^64]: The compiler does so using strict "elision rules", and does not actually peek into the function body to do so.

[^65]: Rust is so strict about this that it will mark code as unreachable and emit illegal instructions if it notices that this is happening.

[^66]: Also, note that strict aliasing does not apply to shared references, either.
`&u32` and `&u64` may alias, but `&mut u32` and `&mut u64` cannot, because a `&mut T` can never alias with any other active reference.

[^67]: To put it in perspective, all references are marked as "dereferenceable" at the LLVM layer, which gives them the same optimization semantics as C++ references.
Merely materializing an invalid reference is Undefined Behavior, because LLVM will treat it as a pointer to valid memory that it can cache reads from and combine writes to.

[^68]: It is common convention in Rust to name the default function for creating a new value `new`; it is not a reserved word.

[^69]: Note the capital S.

[^70]: This is currently fairly restricted; for our purposes, only `Self`, `&Self`, and `&mut Self` are allowed.

> [^71]: There's only aouple of dynamically sized types built into the language; user-defined DSTs exist, but they're a very advanced topic.

Reviewer: ```suggestion
[^71]: There's only a couple of dynamically sized types built into the language; user-defined DSTs exist, but they're a very advanced topic.
```

Author: Done.

Rust string literals are much like C string literals: `"abcd..."`.
Arbitrary ASCII-range bytes can be inserted with `\xNN`, and supports most of the usual escape sequences.
However, all Rust strings are UTF-8 encoded byte slices: `&str` is a wrapper type around `&[u8]` that guarantees that the bytes inside are valid UTF-8.
The type of all string literals is `&'static str`.

Rust string literals can contain arbitrary newlines in them, which can be escaped:
```rust
// Equivalent to "foo\n  bar".
let s = "foo
  bar";
// Equivalent to "foo  bar".
let s = "foo\
  bar";
```

Raw strings disable escape sequences, and are delimited by an arbitrary, matching number of pound signs:
```rust
let s = r"...";
let s = r#" ..."#;
let s = r#####"..."#####;
```
If instead a byte string with no encoding is required, byte strings can be used: `b"..."`.
Their contents must be ASCII (or escaped bytes), and their type is `&[u8]`.
Raw strings can also be byte strings: `br#"..."#`.

Rust also has character literals in the form of `'z'`[^75], though their type is `char`, a 32-bit Unicode code-point.
To get an ASCII byte of type `u8`, instead, use `b'z'`.

### Destructors and RAII {#destructors-and-raii}

> Destctors are special functions that perform cleanup logic when a value has become unreachable (i.e., both the `let` that originally declared it can no longer be named, and the last reference to it expired).

Reviewer 3: Destctors -> Destructors

*   `foo`, an identifier pattern.
    This pattern is exactly like `_`, but it binds the matched value to its name.
    This is the `val` in the `Some(val)` above.
    This can also be used by itself as a default case that wants to do something with the matched-on value.
    The binding can be made mutable by writing `Some(mut val)` instead.
*   Any numeric literal, for a literal pattern.
    This match compares the matched value against the literal value, and doesn't match anything.
    These can also be inclusive ranges: `5..=16`[^86].
*   `(pat1, pat2, /* etc */)`, a tuple pattern.
    This match operates on tuple types, and always succeeds: it extracts the individual elements of a tuple, and applies them to the pattern's subpatterns.
    In particular, the `()` pattern matches the unit value `()`.
```rust
let x: (u32, u32) = ...;
match x {
  (5, u) => { ... },  // Check that first element is five, 
                      // bind the second element to `u`.
  (u, _) => { ... },  // Bind the first element to `u`,
                      // discard the second element.
}

let y: (u32, (u32, u32)) = ...;
match y {
  // All patterns can nest arbitrarily, like expressions.

>   (42, (u, _)) =>  { ... },

Reviewer 3: I'd suggest not using {`...`} to denote code, because it's slightly confusing with `..` in tuple pattern matching. Also, I'd suggest putting `..` up above -- having it as a comment in the code means there are two places you lay out the possibilities. I missed it the first time.

Author: How would you feel about `s#...#/* ... */#g`?

Reviewer 3: I think that would be clearer.

Author: Done.

It is not uncommon for generic wrapper types, which restrict access to a value, to be smart pointers.

Note that, because `Target` is an associated type, a type can only dereference to one other type.

While not quite as relevant to smart pointers, the `Index` and `IndexMut` traits are analogous to the `Deref` and `DerefMut` traits, which enable the `x[...]` subscript syntax.
`Index` looks like this:
```rust
trait Index<Idx> {
  type Output;
  fn index(&self, index: Idx) -> &Self::Output;
}
```

An indexing operation, much like a dereference operation, expands from `x[idx]` into `*(x.index(idx))`.
Note that indexing _can_ be overloaded, and is a useful example of how this overloading through traits can be useful.
For example, `<[u8] as Index<usize>>::Output` is `u8`, while `<[u8] as Index<Range>>::Output` is `[u8]`.
Indexing with a single index produces a single byte, while indexing with a range produces another slice.

### Closures {#closures}

Closures (sometimes called "lambda expressions" in other languages) are function literals that capture some portion of their environment, which can be passed into other functions to customize behavior.

> The classic C equivalent is something like `pthread_create`, which takes a function pointer `void* (*)(void*)` and a `void*`, the latter being "captured state" sent across the thread boundary.

Reviewer 3: I don't think this is a good analogy -- it's not like pthread_create at all. I think it's safe to say that there is no classic C equivalent, but there are in many other languages.

Author: Well, the point I want to get across is that `&dyn Fn` actually behaves that way at the ABI layer. You pass in a function pointer (well, there's an extra vtable indirection...) with some state that's behind a pointer. This is totally a pattern I've seen before; I think qsort() is similar.

I really want an analogy here with a familiar C function, to make it clear that dyn closures are not just a function pointer, but also a pointer to some context.

Reviewer 3: Maybe rather than say "the classic C equivalent" say "the closest analogy in C" and explain how it's different?

Author: I wrote some prose. Tell me what you think!

Reviewer 3: Definitely clearer. My one remaining suggestion relates to this text:

"In a similar way, Rust need extra state to execute, except arg becomes part of the start_routine value. As we'll see, Rust has a number of different ABIs for closures, some of which closely resemble what pthread_create does; in some cases, the function pointer can even be inlined."

I'd suggest that you mention that unlike the void* in pthread_create, where this context must be manually constructed, the Rust language automatically constructs it through static analysis.  I think this stuff is obvious to anyone who's used them before but if a pure C person is reading it this will not be clear. 

I'm touching on this in part because closures are so critical in Tock (e.g., grants, MapCell, etc.).

Finally, I think "Rust need" should be "Rust closures need"?

Author: Done!

This is in contrast to C, where each `.c` file generates a separate object file.[^8] Rust also does not have headers, though it provides a module system for intra-crate code organization that will be discussed later on.
Tock boards are a good example of what a more-complex cargo file looks like: [https://github.com/tock/tock/blob/master/boards/opentitan/Cargo.toml](https://github.com/tock/tock/blob/master/boards/opentitan/Cargo.toml)

Some useful `cargo` subcommands include:
*   `cargo check` runs rustc's checks, but stops before it starts emitting code and optimizing it.
    This is useful for checking for errors during development.
*   `cargo build` builds a library or binary, depending on the crate type.
*   `cargo clippy` runs the Rust linter, Clippy.
*   `cargo doc --open` builds crate documentation and then opens it in the browser.
*   `cargo fmt` runs the Rust formatter[^9].

Also, the contents of the `RUSTFLAGS` environment variable are passed to `rustc`, as a mechanism for injecting flags.

The Rust standard library, like libc, is smaller in embedded environments.

> The standard library consists of three crates: `core`[^10], `alloc`, and `std`.

Reviewer 4: This doesn't visualize correctly in the github markdown. The `[` and `]` disappear.

Author: This Markdown file isn't intended for GH; it's going to show up on the OpenTitan docs website. You can preview it with `util/build_docs.py`.

[Note that this syntax is currently in the process of being redesigned and stablized.](https://blog.rust-lang.org/inside-rust/2020/06/08/new-inline-asm.html)

#### Bit Casting {#bit-casting}

Rust provides a type system trap-door for bitcasting any type to any other type of the same size:
```rust
let x = ...;
let y = std::mem::transmute<A, B>(x);
```
This trap-door is extremely dangerous, and should only be used when `as` casts are insufficient.
The primary embedded-relevant use-case is for summoning function pointers from the aether, since Rust does not allow casting raw pointers into function pointers (since the latter are assumed valid).

[https://doc.rust-lang.org/std/mem/fn.transmute.html](https://doc.rust-lang.org/std/mem/fn.transmute.html) has a list of uses, many of which do not actually require transmute.

#### Linker Shenanigans and Other Attributes {#linker-shenanigans-and-other-attributes}

Below are miscellaneous attributes relevant to embedded programming.
*   `#[link_section = "..."]` is the rust spelling of `__attribute__((section("...")))`, which will stick a symbol in the given ELF (or equivalent) section.

> *   `#[used]` can be used to force a symbol to be kept by the linker[^57] (this is often done in C by marking the symbol as `volatile`).

Author: Added some cavats. Thoughts?

   |
7  |     let handle = new_handle(0x40000);
   |         ------ move occurs because `handle` has type `Handle`,
   |                which does not implement the `Copy` trait
8  |     use_for_scheduling(handle);
   |                        ------ value moved here
9  |     // ...
10 |     use_for_scheduling(handle);
   |                        ^^^^^^ value used here after move
```

Use-after-free errors (and double-free errors) are impossible in Safe Rust.
This particular class of errors (which don't directly involve pointers) are prevented by _move semantics_.
As the error above illustrates, using a variable marks it as having been "moved from": the variable is now an empty slot of uninitialized memory.
The compiler tracks this statically, and compilation fails if you try to move out again.
The variable in which a value is currently stored is said to be its "owner"[^59]; an owner is entitled to hand over ownership to another variable[^60], but may only do so once.

The error also notes that "`Handle` does not implement the `Copy` trait."
Traits proper are a topic for later; all this means right now is that `Handle` has move semantics (the default for new types).

> Types which _do_ implement `Copy` have _copy semantics_; this is how all types in C behave.

Reviewer 5: ```suggestion
Types which _do_ implement `Copy` have _copy semantics_; this is how all primitive types in C behave.
```

Maybe just this, maybe more:

---

I'm afraid this might be a bit misleading, particularly as a document that targets the conceptual model of C programmers. I don't think people think about `&my_struct` as being a "copy semantic of the `my_struct` pointer", rather it's thought of as "passing a reference to `my_struct`". I think it might be worth expounding on this a bit to compare to C notional types rather than just literal types.

Author: I added some hand-waving about "by value" and "by reference" that may or may not be helpful.

Reviewer 5: I think that's much clearer now!

```

Of course, the copy/move distinction is something of a misnomer: reassignment due to copy and move semantics compiles to the same `memcpy` or register move code.
The distinction is purely for static analysis.

### References and Lifetimes {#references-and-lifetimes}

Another class of use-after-free involves stack variables after the stack is destroyed.
Consider the following C:
```c
const int* alloc_int(void) {
  int x = 0;
  return &x;
}
```
This function is obviously wrong, but such bugs, where a pointer outlives the data it points to, are as insidious as they are common in C.

Rust's primary pointer type, _references_, make this impossible.
References are like raw pointers, except they are always well-aligned, non-null, and point to valid memory; they also have stronger aliasing restrictions than C pointers.
Let's explore how Rust achieves this last guarantee.

Consider the following Rust program:
```rust
fn alloc_int() -> &i32 {
  let x = 0i32;
  &x
}
```
This program will fail to compile with a cryptic error: `missing lifetime specifier`.
Clearly, we're missing something, but at least the compiler didn't let this obviously wrong program through.

> A lifetime, denoted by a symbol like `'a`[^61] (the apostrophe is often pronounced "tick"), represents a contiguous region of source code[^62].

Reviewer 5: ```suggestion
A lifetime, denoted by a symbol like `'a`[^61] (the apostrophe is often pronounced "tick"), labels a region of code[^62].
```

The full reality may be too complex for this point in this document, but lifetimes are not a 1:1 mapping to contiguous blocks of code, the nomicon gives examples of more complex lifetimes: https://doc.rust-lang.org/nomicon/lifetimes.html

Author: Lifetimes *are* contiguous, though reborrowing (e.g., temporarially disabling a mutable reference by creating another reference out of it with a shorter lifetime attached) can make it appear as though they have holes in them. In particular, it only every possible to name a lifetime `'a` within the region named by `'a`, which is the reason why there is no "empty" lifetime (i.e., the complement of `'statc`).

It can be easier to think of lifetimes not in terms of their "physical shape", but rather in terms of what references are active within them (even if those references are "pinned" by reborrows).

That said, I'll drop the contiguous label to minimize confusion.

Lifetimes, much like types, do not exist at runtime; they only exist for the compiler to perform _borrow checking_, in which the compiler ensures that references only exist within their respective lifetimes.
A special lifetime, `'static` represents the entire program.
It is the lifetime of constants and global variables.

Consider the following Rust code:
```rust
let x: i32 = 42;
let y: &'a i32 = &x;  // Start of 'a.
use_reference(y);
use_value(x);  // End of 'a, because x has been moved.
use_reference(y);  // Error: use of y outside of 'a.
```
Reference lifetimes start when the reference is taken, and end either when the lifetime goes out of scope or when the value referenced is moved.
Trying to use the reference outside of the lifetime is an error, since it is now a dangling pointer.

Rust often refers to references as _borrows_: a reference can borrow a value from its owner for a limited time (the lifetime), but must return it before the owner gives the value up to someone else.
It is also possible for a reference to be a borrow of a borrow, or a _reborrow_: it is always possible to create a reference with a shorter lifetime but with the same value as another one.

> Reborrowing is usually performed implicitly by the compiler, usually around call sites, but can be performed explicitly by writing `&*x`.

Reviewer 5: Maybe add an example where reborrowing is necessary? I just went hunting for places where we use `&*` in Tock, and the first one I tried compiled fine with the reborrow removed..., this might be something that's unintentionally overused at the moment?

Reviewer 5: Further update: https://github.com/tock/tock/pull/1973

Author: I've definitely come across examples of this... but I'm having trouble thinking of any on the spot. I think calling out the syntax isn't a bad idea, since people may see it and be confused (since it's meaningless in C and probably a bad idea in C++).

Traits themselves, however, are _not_ types.

A very simple trait from the standard library is `Clone`[^88]:
```rust
trait Clone {
  fn clone(&self) -> Self;
}
```

A type satisfying `Clone`'s interface (in Rust parlance, "implements `Clone`") has a `clone` method with the given signature, which returns a duplicate of `self`.
To implement a trait, you use a slightly funny `impl` syntax[^89]:
```rust
impl Clone for MyType {
  fn clone(&self) -> Self { /* implementation */ }
}
```

This gives us a consistent way to spell "I want a duplicate of this value".
The standard library provides traits for a number of similar operations, such as `Default`[^90], for providing a default value, `PartialEq`[^91] and `Eq`, for equality, `PartialOrd`[^92] and `Ord`, for ordering, and `Hash`[^93], for non-cryptographic hashing.

The above traits are special in that they have trivial implementations for a struct or enum, assuming that all fields of that struct or enum implement it.
The `#[derive()]` syntax described in the "Ownership" section can be used with any of these traits to automatically implement them for a type.

> It is not uncommon for POD types to look like this:

Reviewer 5: ```suggestion
It is not uncommon for Plain-Old-Data (POD) types to look like this:
```

Author: Done.

```rust
#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]

> struct MyPod { ... }

Reviewer 5: ```suggestion
struct MyPod {
    a: u8,
    b: u8,
    // x: OtherCustomStruct,    <-- The derive would fail as the compiler cannot automatically implement `Copy` and related functions for non-POD types 
}
```

Author: Done.

For example, the standard library trait `Sync` is implemented by all types which synchronize access[^98].
Because the invariant this trait asserts is beyond what the compiler can check, it is an unsafe trait.

Trait methods may separately be marked as `unsafe`.
This is usually done to indicate that not only does care need to be taken in implementing the trait, but calling the function also requires care (and uttering `unsafe`).
This is separate from marking a trait as `unsafe`, and it is not necessary to mark a trait as `unsafe` for it to have `unsafe` methods.

#### Auto Traits

Auto traits are a compiler mechanism for automatically implementing certain traits; in the standard library's source code, this shows up as `auto trait Foo {}` (though this syntax is unavailable for normal libraries).
Auto traits are implemented automatically for a struct or enum type if all of its fields also implement that trait, and are used for exposing transitive properties to the trait system.
For example, `Send` and `Sync` are auto traits; a number of other marker traits[^99] are also auto traits.

> 

Reviewer 5: Perhaps something that explains why some things are Auto Traits versus requiring a manual Derive?

Author: Done.

### Generics {#generics}

_Generic programming_ is writing source code that can be compiled for many types.
Generics are one of Rust's core features, which enable polymorphic[^100] static dispatch.

Functions can be made generic by introducing _type parameters_, using a syntax similar to explicit lifetimes[^101]:
```rust
fn identity<T>(x: T) -> T {
  x
}
```
This function accepts a value of any type and immediately returns it.
It can then be called like this: `identity::<i32>(42)`[^102].
Using a generic function with all of its type parameters filled in causes it to be _instantiated_ (or _monomorphized_), resulting in code being generated for it.
This process essentially consists of replacing each occurrence of `T` with its concrete value.

Each distinct instantiation is a separate function at runtime, with a separate address, though for functions which generate identical code, like `identity::<i32>` and `identity::<u32>`, the linker may deduplicate them.
Overzealous use of generic code can lead to binary bloat.

> Most of the time, the `::<>` bit is unnecessary, since Rust type deduction can infer it: `let x: u64 = identity(42);` will infer that `T = u64`.

Reviewer 5: ```suggestion
Most of the time, the `::<>` bit (called the `turbofish` in Rust parlance) is unnecessary, since Rust type deduction can infer it: `let x: u64 = identity(42);` will infer that `T = u64`.
```

[^3]: Alternative implementations exist, though `rustc` is the reference implementation.
Rust is still in the process of being specified, so, for now, `rustc`'s behavior is mostly normative.

[^4]: Not every project uses rustup, but it's frequently necessary for "nightly" features.

[^5]: Inline assembly is the most salient of these.
Nightly also provides support for the sanitizers, such as ASAN and TSAN.

[^6]: Example from Tock: [https://github.com/tock/tock/blob/master/rust-toolchain](https://github.com/tock/tock/blob/master/rust-toolchain)

[^7]: Rust libraries use their own special "rlib" format, which carries extra metadata past what a normal .a file would.

[^8]: While each crate is essentially a giant object file, Rust will split up large crates into smaller translation units to aid compilation time.

[^9]: The formatter is kind of a new component, so may require separate installation through `rustup`.
See [https://github.com/rust-lang/rustfmt#on-the-stable-toolchain](https://github.com/rust-lang/rustfmt#on-the-stable-toolchain).

[^10]: [https://doc.rust-lang.org/core/index.html](https://doc.rust-lang.org/core/index.html)

> 

Reviewer 4: For some reason footnote [^10] disappears in the github rendered version of this markdown file.  https://github.com/lowRISC/opentitan/blob/e52d7b694c00814deaef007c9ba6677f66b4066d/doc/ug/rust_for_c.md#footnotes

Maybe this isn't a concern since the footnotes don't work properly in github markdown anyways?

This is because Rust has a separate boolean type, for which `!x` is logical not.

[^13]: Rust has slightly less absurd precedence rules around bitwise operators; `x ^ y == z` is `(x ^ y) == z`.

[^14]: In C, signed overflow is UB, and unsigned overflow always wraps around.

[^15]: Really, this causes a "panic", which triggers unwinding of the stack.
In embedded contexts, unwinding is disabled, so this is just a jump to the panic handler, which performs a platform-specific abort.

[^16]: rustc performs the former in debug builds and the latter in release builds.

[^17]: Creating `bool`s with any other representation, using Unsafe Rust, is instant UB.
This is sometimes called the "validity invariant", and applies to other restricted-range types, such as padding bytes in structs.
["What The Hardware Does" is not What Your Program Does: Uninitialized Memory](https://www.ralfj.de/blog/2019/07/14/uninit.html) provides a short explanation about why this is important.

[^18]: These functions are frequently entry-points for LLVM intrinsics, similar to GCC/Clangs `__builtin_clz()` and similar.

[^19]: This syntax: `(my_struct_t) { .a = 0, .b = 1 }`.

> [^20]: Rust allows trailing commas basically everywhere.

Reviewer 4: Something is screwed up around footnote 19/20. Either could successfully be applied where [^19] is in the text, but afterwards all the remaining footnotes are off by one. For example [^20] in the text should really be [^21].

Reviewer 4: Still pretty sure your footnotes are off-by-one starting at [^20]. I don't think this is just a github visualization issue.

[^32]: Or by casting zero.

[^33]: Rust currently does not have an equivalent of the `->` operator, though it may get one some day.
As such, it is not possible to get pointers to fields of a pointer-to-struct without creating a reference along the way.

[^34]: We will get into references later.

[^35]: We will learn more about "move semantics" when we discuss ownership.

[^36]: However, great care should be taken when using these methods on types that track ownership of a resource, since this can result in a double-free when the pointee is freed.

[^37]: [https://doc.rust-lang.org/std/primitive.pointer.html#method.read_unaligned](https://doc.rust-lang.org/std/primitive.pointer.html#method.read_unaligned)

[^38]: [https://doc.rust-lang.org/std/primitive.pointer.html#method.write_unaligned](https://doc.rust-lang.org/std/primitive.pointer.html#method.write_unaligned)

[^39]: These are effectively memcpys: they will behave as if they read each byte individually with no respect for alignment.

[^40]: [https://doc.rust-lang.org/std/primitive.pointer.html#method.copy_to](https://doc.rust-lang.org/std/primitive.pointer.html#method.copy_to)

> [^41]: [https://doc.rust-lang.org/std/primitive.pointer.html#method.copy_to_nonoverlapping](https://doc.rust-lang.org/std/primitive.pointer.html#method.copy_to_nonoverlapping)

Reviewer 4: 40 and 41 are also messed up where they are referenced in the text and don't appear at the end in the visualization on github. Maybe it's because they are only links with no text?

Reviewer 4: I'm pretty confused why this is occurring if github markdown doesn't support footnotes...

Reviewer 4: There are more (such as 56 and 57 below) but I'm going to stop listing them because 1) I'm definitely going to miss some anyways and 2) I'm not sure if it's a bug you care about.

```

Of course, the copy/move distinction is something of a misnomer: reassignment due to copy and move semantics compiles to the same `memcpy` or register move code.
The distinction is purely for static analysis.

### References and Lifetimes {#references-and-lifetimes}

Another class of use-after-free involves stack variables after the stack is destroyed.
Consider the following C:
```c
const int* alloc_int(void) {
  int x = 0;
  return &x;
}
```
This function is obviously wrong, but such bugs, where a pointer outlives the data it points to, are as insidious as they are common in C.

Rust's primary pointer type, _references_, make this impossible.
References are like raw pointers, except they are always well-aligned, non-null, and point to valid memory; they also have stronger aliasing restrictions than C pointers.
Let's explore how Rust achieves this last guarantee.

Consider the following Rust program:
```rust
fn alloc_int() -> &i32 {
  let x = 0i32;
  &x
}
```
This program will fail to compile with a cryptic error: `missing lifetime specifier`.
Clearly, we're missing something, but at least the compiler didn't let this obviously wrong program through.

> A lifetime, denoted by a symbol like `'a`[^61] (the apostrophe is often pronounced "tick"), labels region of source code[^62].

Reviewer 5: ```suggestion
A lifetime, denoted by a symbol like `'a`[^61] (the apostrophe is often pronounced "tick"), labels a region of source code[^62].
```

Author: Done.

In general, a few rules of thumb are useful:
*   Every `unsafe fn` should declare, in documentation, what invariants it assumes that the caller will uphold, and what state it will leave the machine in.
    For example, `<[T]>::get_unchecked(n)` elides the bounds check for the indexing operation, and it is up to the caller to uphold it instead.
*   Every time `unsafe` code calls into a non-`unsafe` function, it must ensure that no violated invariants, which could trigger Undefined Behavior, are observable in that safe code.
    For example, if we have a type that maintains the invariant that `len > 0`, and we temporarily set it to `0` during an `unsafe` block, it must be restored to `> 0` before any safe methods can be called on that type[^128].
*   Unsafe code should be kept to the absolute minimum, and wrapped in safe interfaces that assert invariants, either through static type-system guarantees or through runtime checks.
    Every line of unsafe code is a place where the engineering cost of Rust's guarantees are wasted.

In other words, Safe Rust is able to freely assume that Rust's guarantees on aliasing, ownership, and representation of values hold at all times.

> This assumption is perversive: not only does the compiler use it aggressively optimize code for speed and size, but other library code, such as the destructors of wrapper types, assume it, too.

Reviewer: Did you mean "pervasive"? "perversive" is an esoteric word choice here.

Author: I can't English.

However, effective use of such enums requires _pattern matching_, which requires its own section to explain.
We'll see these enums again when we discuss patterns.

Just like with structs, `#[derive]` can be used on enums to define comparison operators, which are defined analogously to the struct case.

#### Arrays {#arrays}

Rust arrays are just C arrays: inline storage of a compile-time-known number of values.
`T[N]` in C is spelled `[T; N]` in Rust.
Arrays are created with `[a, b, c]` syntax, and an array with lots of copies of the same value can be created using `[0x55u8; 1024]`[^26].
A multi-dimensional array can be declared as an array of arrays: `[[T; N]; M]`.

Array elements can be accessed with `x[index]`, exactly like in C.
Note, however, that Rust automatically inserts bounds checks around every array access[^27]; failing the bounds check will trigger a panic in the program.
Unsafe Rust can be used to cheat the bounds check when it is known (to the programmer, not rustc!) to be unnecessary to perform it, but when it is performance critical to elide it.

> Rust arrays are "real" types, unlike in C. They can be passed by value intofunctions, and returned by value from functions.

Reviewer 6: `intofunctions` -> `into functions`

### Analogues of other functionality {#analogues-of-other-functionality}

#### Volatile {#volatile}

Rust does not have a `volatile` qualifier.
Instead, volatile reads can be performed using the `read_volatile()`[^55] and `write_volatile()`[^56] methods on pointers, which behave exactly like volatile pointer dereference in C.

Note that these methods work on types wider than the architecture's volatile loads and stores, which will expand into a series of volatile accesses, so beware.
The same caveat applies in C: `volatile uint64_t` will emit multiple accesses on a 32-bit machine.

#### Inline Assembly {#inline-assembly}

Rust does not quite support inline assembly yet.
Clang's inline assembly syntax is available behind the unstable macro `llvm_asm!()`, which will eventually be replaced with a Rust-specific syntax that better integrates with the language.
`global_asm!()` is the same, but usable in global scope, for defining whole functions.
Naked functions can be created using `#[naked]`. See [https://doc.rust-lang.org/1.8.0/book/inline-assembly.html](https://doc.rust-lang.org/1.8.0/book/inline-assembly.html).

> [Note that this syntax is currently in the process of being redesigned and stablized.](https://blog.rust-lang.org/inside-rust/2020/06/08/new-inline-asm.html)

Reviewer 6: `stablized` -> `stabilized`

This means more optimization opportunities, without safe code needing to do anything.[^65]

Finally, it should go without saying that references are only useful for main memory; Rust is entitled to generate spurious loads and stores for (possibly unused!) references[^66], so MMIO should be performed exclusively through raw pointers.

[https://doc.rust-lang.org/book/ch10-03-lifetime-syntax.html](https://doc.rust-lang.org/book/ch10-03-lifetime-syntax.html) elaborates further on the various lifetime rules.

#### Operations on References {#operations-on-references}

Rust references behave much more like scalar values than like pointers (borrow-checking aside).
Because it is statically known that every reference, at all times, points to a valid, initialized value of type `T`, explicit dereferencing is elided most of the time (though, when necessary, they can be dereferenced: `*x` is an lvalue that can be assigned to).

Rust does not have a `->` operator, but, for `x: &T`, the dot operator behaves as if `x` were a `T`.
If `field` is a field of `T`, for example, `x.field` is the lvalue of `field` (which would be spelled `x->field` in C).

> This applies even for heavily nested rerences: the dot operator on `&&&T` will trigger three memory lookups.

Reviewer 6: `rerences` -> `references`

Let's unpack the syntax.

Functions in an `impl` block are called "inherent functions" or "methods", depending on whether they take a `self` parameter.
Inherent functions don't take a `self`, and are called like `Counter::new()`[^67].

A `self` parameter is a parameter named `self` (which is a keyword) and having a type involving `Self`[^68] (another keyword, which is a type alias for the type the `impl` block is for), such as `&Self`[^69].
The syntax `self`, `&self`, `mut self`, and `&mut self` are sugar for the common forms `self: Self`, `self: &Self`, `mut self: Self` and `self: &mut Self`, representing self-by-value, self-by-reference, self-by-mut-value, and self-by-mut-reference.

Methods can, thus, be called like so: `my_counter.inc()`.
Methods are really just normal functions: you could also have called this like `Counter::inc(&mut my_counter)`.
Note that calling a function that takes `&self` or `&mut self` triggers a borrow of the receiving type; if a `&self` function is called on a non-reference value, the value will have its address taken, which gets passed into the method.

> `impl` blocks, like other items, can be parametrized by lifetimes.

Reviewer 6: `parametrized` -> `parameterized`

The above traits are special in that they have trivial implementations for a struct or enum, assuming that all fields of that struct or enum implement it.
The `#[derive()]` syntax described in the "Ownership" section can be used with any of these traits to automatically implement them for a type.
It is not uncommon for Plain Old Data (POD) types to look like this:
```rust
#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
struct MyPod {
  pub a: u8,
  pub b: u8,
  // The following line wouldn't compile, because `derive(Eq)` requires
  // all fields to be `Eq`.
  // c: NonEq,
}
```

Traits can also provide built-in methods implemented in terms of other methods, to provide a default implementation (which can be overridden if a more efficient one is available for a particular type).
The full `Clone` trait actually looks like this:
```rust
pub trait Clone {
  fn clone(&self) -> Self;
  fn clone_from(&mut self, source: &Self) {
    *self = source.clone();
  }
}
```

> Implementors don't need to provide `clone_from`, but are allowed to do so if the default implementation isn't good enough.

Reviewer 6: `Implementors` -> `Implementers`

[^36]: However, great care should be taken when using these methods on types that track ownership of a resource, since this can result in a double-free when the pointee is freed.

[^37]: [https://doc.rust-lang.org/std/primitive.pointer.html#method.read_unaligned](https://doc.rust-lang.org/std/primitive.pointer.html#method.read_unaligned)

[^38]: [https://doc.rust-lang.org/std/primitive.pointer.html#method.write_unaligned](https://doc.rust-lang.org/std/primitive.pointer.html#method.write_unaligned)

[^39]: These are effectively memcpys: they will behave as if they read each byte individually with no respect for alignment.

[^40]: [https://doc.rust-lang.org/std/primitive.pointer.html#method.copy_to](https://doc.rust-lang.org/std/primitive.pointer.html#method.copy_to)

[^41]: [https://doc.rust-lang.org/std/primitive.pointer.html#method.copy_to_nonoverlapping](https://doc.rust-lang.org/std/primitive.pointer.html#method.copy_to_nonoverlapping)

[^42]: Rust also provides an abstraction for dealing with uninitialized memory that has somewhat fewer sharp edges: [https://doc.rust-lang.org/std/mem/union.MaybeUninit.html](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html).

> [^43]: Rust also allows taking the addresses of rvalues, which simply shoves them onto .rodata or the stack, depending on mutation.

Reviewer 6: Backticks around `.rodata`

> For example, `&0` is valid, and produces an address in .rodata, while `&mut 0` will push into the stack.

Reviewer 6: Same here.
